### PR TITLE
Debugging TOL-10M dataset reproduction process

### PR DESCRIPTION
**Pull request by thompsonmj**

### CODE REVIEW

**Comment by egrace479** [comment]
@thompsonmj please check the version I updated. My copy is still running on OSC, but it does seem to be working now.

File: slurm/make-dataset-wds_reproduce.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+module load miniconda3/24.1.2-py310
+conda --version
+which conda
+
+conda activate /users/PAS2136/thompsonmj/.conda/envs/bioclip-train
+pip install -e /fs/scratch/PAS2136/thompsonmj/bioclip
+which python
+whoami
+
+~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag v3.3-cross-entropy --split val --workers $SLURM_CPUS_PER_TASK
```
Comment (egrace479) [inline slurm/make-dataset-wds_reproduce.sh:24]:
```suggestion
~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag CVPR-2024 --split val --workers $SLURM_CPUS_PER_TASK
```

File: slurm/make-dataset-wds_reproduce.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag v3.3-cross-entropy --split train_small --workers $SLURM_CPUS_PER_TASK
```
Comment (egrace479) [inline slurm/make-dataset-wds_reproduce.sh:25]:
```suggestion
~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag CVPR-2024 --split train_small --workers $SLURM_CPUS_PER_TASK
```

File: slurm/make-dataset-wds_reproduce.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag v3.3-cross-entropy --split train --workers $SLURM_CPUS_PER_TASK
```
Comment (egrace479) [inline slurm/make-dataset-wds_reproduce.sh:26]:
```suggestion
~/.conda/envs/bioclip-train/bin/python scripts/evobio10m/make_wds_reproduce.py --tag CVPR-2024 --split train --workers $SLURM_CPUS_PER_TASK
```

**Review by egrace479 — COMMENTED** [review]
Set tag in folder name to match the version (release) of the data.

**Comment by hlapp** [comment]
BTW is there a reason we're invoking the python interpreter through its path semi-hardcoded and into the conda environment, as opposed to simply activating the conda environment first (which presumably needs to be done anyway)?

As an aside, conda environments can get quite big in terms of storage they take up, and it's therefore generally not a good idea to place them into one's home directory on a shared HPC. (Home directories on shared HPCs typically have quite a limited storage quota, and group or project directories is normally where a lot more storage is available or at least where it can be added on demand.)

File: slurm/download_bioscan.slurm [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# Extract BIOSCAN archive
+zip_archive="${bioscan_root}/cropped_256.zip"
+extract_bioscan "$zip_archive" "$bioscan_root"
+
+# End timing
+time_end=$(date +%s)
+time_elapsed=$((time_end - time_start))
+
+# Notify completion
+echo "Total execution time: $time_elapsed seconds" > "${logs_path}/download_bioscan_completed.txt"
```
Comment (egrace479) [inline slurm/download_bioscan.slurm:24]:
```suggestion
echo "Total execution time: $time_elapsed seconds" > "${logs_path}/download_bioscan_completed.txt"

```

File: scripts/setup_download_tol-10m_components.bash [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+# Network-intenstive download of 63x ~30GB .tar.gz files with EOL images (~2.5h on OSC, 200-300MiB/s)
+# Wall time may need to increase if `wget` is used instead of `aria2c`
+EOL_SLURM_TIME="18:00:00"
+EOL_SLURM_NTASKS_PER_NODE=12
+EOL_SLURM_OUTPUT="${REPO_ROOT}/logs/download_eol_%j.out"
+EOL_SLURM_ERROR="${REPO_ROOT}/logs/download_eol_%j.err"
+
+# Network intensive download (~11min on OSC, 340MiB/s), CPU-bound extraction (~88min total on OSC)
+INAT21_SLURM_TIME="0:10:00"
```
Comment (thompsonmj) [inline scripts/setup_download_tol-10m_components.bash:44]:
```suggestion
INAT21_SLURM_TIME="8:00:00"
```

File: slurm/download_eol.slurm [hunk 6]
```diff
@@ -4,12 +4,12 @@ set -ex
...
 # Source setup script
-source "$SLURM_SUBMIT_DIR/setup_download_tol-10m_components.bash"
+source "$SLURM_SUBMIT_DIR/scripts/setup_download_tol-10m_components.bash"
 
-# source "../scripts/setup_download_tol-10m_components.bash" # Use this for troubleshooting or running this component independently
+# source "scripts/setup_download_tol-10m_components.bash" # Use this for troubleshooting or running this component independently
 
 # Download EOL data
-for i in $(seq -w 01 63); do
+for i in $(seq -w 01 03); do
```
Comment (egrace479) [inline slurm/download_eol.slurm:12]: Why did this change?

**Comment by egrace479** [comment]
@samuelstevens, this should work as described in updated directions in PR #14. If you encounter any issues or something seems off, please make suggestions there too.

**Comment by penfever** [comment]
@samuelstevens @thompsonmj @egrace479, just curious, has this bug been resolved by this PR?

**Comment by thompsonmj** [comment]
Apologies for the delay @penfever. There is still a lingering issue producing the catalog where common names are not reproduced as expected. We hope to resolve it within the next week.

**Comment by thompsonmj** [comment]
Issue with null common names resolved.

**Comment by thompsonmj** [comment]
@penfever, I believe this completes the fixes needed to address #13. 

Could you please attempt to build the dataset from [this branch](https://github.com/Imageomics/bioclip/tree/reproduce_tol10m) following [these instructions](https://github.com/Imageomics/bioclip/blob/docs/reconstruct/docs/imageomics/treeoflife10m.md#how-to-create-treeoflife-10m)?

Please reach out if there are additional questions and also if you are able to complete the reconstruction. It will be helpful to know that it can work in someone else's hands.

**Comment by egrace479** [comment]
@thompsonmj, following this step in [our directions](https://github.com/Imageomics/bioclip/blob/docs/reconstruct/docs/imageomics/treeoflife10m.md):
```
conda env create -f requirements-training.yml --solver=libmamba -y
conda activate bioclip-train
pip install -e .
```
we wind up uninstalling and re-installing a bunch of packages like `torch` and `huggingface_hub` to get older versions (those listed in [requirements.txt](requirements.txt) instead of the ones in [requirements-training.yml](requirements-training.yml). Seems potentially redundant to do both. Should we instead have a `tol` environment to do `pip install -e .` in?

**Review by egrace479 — APPROVED** [review]
Reproduction seems to be working on our end. We may want to consider cleaning up the environment in the future to reduce redundancy.

**Comment by thompsonmj** [comment]
> ... we wind up uninstalling and re-installing a bunch of packages ...

> We may want to consider cleaning up the environment ...

Yes, we can address this in another PR. 

> Reproduction seems to be working on our end.

Right, based on my latest full dataset reproduction test run, all splits end up as expected.

**Pull request merged**